### PR TITLE
Upgrade aws-sdk and switch to S3 virtual host style URLs

### LINKED ISSUE #511: Amazon signature version update breaks ofn
Amazon is changing its signature protocol, starting with newest servers but eventually all.

That means Frankfurt S3 wont work with openfoodnetwork as it's currently setup, and needs signature v4, whatever that means in practise.

http://stackoverflow.com/questions/27847888/aws-s3-in-rails-how-to-set-the-s3-signature-version-parameter

### LINKED ISSUE #4177: Amazon S3 Path Deprecation - URL for buckets have changed
## Description
Amazon are changing the URL for buckets from https://s3.amazonaws.com/bucketname/ to https://bucketname.s3.amazon.com/

I'm not sure if this is a gradual roll out, as my bucket is brand new it may be that the old path-style URL has been disabled only for new accounts. The information I could find was that it is being deprecated by September 30 2020.

https://aws.amazon.com/blogs/aws/amazon-s3-path-deprecation-plan-the-rest-of-the-story/

## Context
Any image uploaded to the S3 bucket is broken due to incorrect URL (might only be for new buckets).

## Severity
bug-s2: a non-critical feature is broken, no workaround

## Possible Fix
Change URL to new virtual-host style path or give option to switch to it.

### PR DESCRIPTION
**Pull request by luisramos0**
#### What? Why?

Closes #511 
Closes #4177 

Here I re-use the existing attachment URL variable. **I dont think attachment_url is used right now, only attachment_path is taken into account as far as I understand**, at least with images on S3.
Have a look at the config change PR to see how it will look like:
https://github.com/openfoodfoundation/ofn-install/pull/667

#### What should we test?
We need to deploy this and verify the existing images are still working as well as verify new uploads.
There's very little risk this will affect non-s3 images setup.
We should test that this will work with images stored in a region outside the USA, I have tested locally with eu-west-3 and it worked well.

We need to verify that this version of aws-sdk keeps db backups to S3 working. We need to setup backups to S3 in staging before we stage this PR and then see if backups are still correctly taken after this PR is staged.

#### Release notes
Changelog Category: Technical changes
Upgraded AWS SDK used to interac with AWS S3 and backups.
We can now use all S3 buckets to store images.

#### Dependencies
Before the release with this PR is deployed, this PR with new URLs for all servers needs to be merged https://github.com/openfoodfoundation/ofn-install/pull/667 and the servers provisioned.

### CODE REVIEW

**Review by Matt-Yorkley — APPROVED** [review]

**Review by andrewpbrett — APPROVED** [review]
This looks fine to me in theory 😄 I think the testing plan makes sense too. I'll move to Dev Test Ready.

File: app/models/spree/image.rb [hunk 1]
```diff
@@ -1,5 +1,7 @@
 # frozen_string_literal: true
 
+require 'spree/core/s3_support'
```
Comment (Matt-Yorkley) [inline app/models/spree/image.rb:3]: I think we might need this line in Bye Bye Spree...

**Comment by Matt-Yorkley** [comment]
Rails 4.2 also requires this `aws-sdk` version bump before the upgrade can start :+1:

File: app/models/spree/image.rb (hunk 1 above)
Comment (luisramos0) [reply app/models/spree/image.rb:3]: yesterday when trying bootboot I had to remove something like this to fix a spec, some error with loading this concern multiple times in Enterprise:
https://github.com/openfoodfoundation/openfoodnetwork/pull/6298/files#diff-31c82445b1c4ad30a20486262f1a746b1b2212baf265998d418d043ad005ce15L3

Maybe a mix of bye bye spree and rails 4.1 upgrade (I think auto class loading did change from 4.0 to 4.1).

**Comment by Matt-Yorkley** [comment]
Deployment failed at the migration step with this:
```
rake aborted!
NoMethodError: undefined method `set_attachment_attributes' for #<Class:0x0000557c21ac4a98>
Did you mean?  set_attachment_attribute
set_s3_attachment_attributes
/home/openfoodnetwork/.gem/ruby/2.3.0/gems/activerecord-4.0.13/lib/active_record/dynamic_matchers.rb:22:in `method_missing'
/home/openfoodnetwork/apps/openfoodnetwork/releases-old/2020-11-09-222214/config/initializers/spree.rb:40:in `<top (required)>'
/home/openfoodnetwork/apps/openfoodnetwork/releases-old/2020-11-09-222214/config/initializers/db2fog.rb:1:in `require_relative'
/home/openfoodnetwork/apps/openfoodnetwork/releases-old/2020-11-09-222214/config/initializers/db2fog.rb:1:in
...etc
```

**Comment by Matt-Yorkley** [comment]
It failed via Semaphore as well... :thinking:

**Comment by Matt-Yorkley** [comment]
Okay the PR is now fixed, but provisioning `fr-staging` with https://github.com/openfoodfoundation/ofn-install/pull/667 and then deploying this breaks all images...

**Comment by luisramos0** [comment]
did you check images were working in staging FR before the PR? https://github.com/openfoodfoundation/openfoodnetwork/issues/6290

**Comment by RachL** [comment]
@Matt-Yorkley @luisramos0 staging FR images have been brocken for a while : https://github.com/openfoodfoundation/openfoodnetwork/issues/6290

**Comment by Matt-Yorkley** [comment]
Ah... were they all broken already? I didn't check before :see_no_evil: 

I'll re-test it on another server.

**Comment by Matt-Yorkley** [comment]
After switching back to `master` and provisioning `fr-staging` again, the images are not broken...

I'll do some more testing but currently i think this can't be merged.

**Comment by luisramos0** [comment]
s3 is not configured in ofn-secrets staging FR. It's working with files stored on the disk.
I'd configure fr-staging to point to s3 (i think there's a bucket in aws already for staging), see the images are working and then try it with this PR :+1:

**Comment by Matt-Yorkley** [comment]
Yeah I'm just looking at it now. Images are broken on `au-staging` as well, it's the only other one that uses S3. :see_no_evil:

**Comment by Matt-Yorkley** [comment]
> think there's a bucket in aws already for staging

I'll check the git history in `ofn-secrets` :+1:

**Comment by luisramos0** [comment]
maybe easier to use the aws credentials in bitbucket to get the api key again.

**Comment by Matt-Yorkley** [comment]
Ok, I set the credentials and correct bucket on `fr-staging` but I'm getting this error on image upload, which is the same one seen previously in Bugsnag: 
```
AWS::S3::Errors::InvalidRequest The authorization mechanism you have provided is not supported. Please use AWS4-HMAC-SHA256.
```

**Comment by luisramos0** [comment]
right, that's what you are seeing with master connecting to the bucket in Europe? That would be good because there would be 2 solutions to that problem:
- use a bucket in us-east
- use this PR :-)

**Comment by luisramos0** [comment]
oh, also, when using/testing this PR you need this PR for ofn-install! https://github.com/openfoodfoundation/ofn-install/pull/667/files

**Comment by Matt-Yorkley** [comment]
Yeah, I've been using 667 and provisioning :+1:

**Comment by Matt-Yorkley** [comment]
Ok, after finally sorting out S3 configs on `fr-staging` and a lot of messing around, this is what the new image URLs look like:

Screenshot from 2020-11-18 23-51-39

Success!

**Comment by luisramos0** [comment]
Great news!

**Comment by RachL** [comment]
@Matt-Yorkley do you know if this is enough to change a bucket's region or if I should open a new issue?

**Comment by luisramos0** [comment]
We need to copy the images to the bucket in the new region before we switch.
 I am not sure how easy it is to copy the images over.

**Comment by RachL** [comment]
@luisramos0 not only images, backups as well. Actually the backups are the more annoying thing for us to leave in an US region. I don't know if it is tight to the system for images?
I have opened a new issue here: https://github.com/openfoodfoundation/ofn-install/issues/672

**Comment by luisramos0** [comment]
Ok. Images and backups use the same s3 api keys but the region can be
different i think. I dont think we needed this issue to move the backups.
Moving backups i think is way easier than moving the images. Maybe two
issues so we can move the backups immediately? (there are a few umverified
assumptions in what i am saying here...)

**Pull request merged**